boa/precompile.py

```diff
@@ -0,0 +1,122 @@
+    typs = [x.split(":")[1].strip().split("[")[0].lower() for x in args]
+    names = [x.split(":")[0].strip().split("[")[0].lower() for x in args]
+
+    # Combine the cleaned argument types and return them in parentheses
+    return f"({','.join(typs)})", names
+
+def parse_address_string(address: str | bytes) -> bytes:
+    # Convert address string to bytes
+    if isinstance(address, str):
+        if address.startswith("0x"):
+            address = address[2:]
+        return bytes.fromhex(address)
+    elif isinstance(address, bytes):
+        return binascii.unhexlify(bytes(address.hex(), "utf-8").zfill(40))
+
+class PrecompileBuiltin(BuiltinFunction):
+    # id can be name
+    _id = ""
+    # list of inputs (parsed from signature)
+    _inputs = []
+    # return type (parsed from signature)
+    _return_type = None
+    _address = None
+
+    def __init__(self, name: str, inputs, return_type, address):
+        self._id = name
+        self._inputs = inputs
+        self._return_type = return_type
+        self._address = address
+
+    @process_inputs
+    def build_IR(self, expr, args, kwargs, context):
+        # allocate buffer for data to pass to precompile
+        # we will pass 4-byte function selector and 32-byte argument
+        buf = context.new_internal_variable(self._return_type)
+
+        args_as_tuple = ir_tuple_from_args(args)
+        args_abi_t = args_as_tuple.typ.abi_type
+
+        ret = ["seq"]
+
+        # store byte selector
+        ret += [["mstore", buf, 0x12341234]]
+
+        # store abi-encoded argument at buf+4
+        length = abi_encode(buf + 4, args_as_tuple, context, args_abi_t.size_bound(), returns_len=True)
+
+        precompile_address_str = "0x" + str(self._address.hex().zfill(40))
+        precompile_address = int(precompile_address_str, 16)
+
+        # call precompile
+        ret.append(["staticcall",
+                    "gas", # fwd all gas
+                    precompile_address, # precompile address
+                    buf, # argsOffset
+                    add_ofst(length, 4), # abi-encoded length + byte selector
+                    buf, # overwrite argsOffset with result
+                    self._return_type.abi_type.size_bound() # return length
+                    ])
+        ret += [buf]
+
+        return IRnode.from_list(ret, typ=self._return_type, location=MEMORY)
+
+def precompile(signature: str, address: Optional[str | bytes] = None):
+    def decorator(func):
+        arg_types, names = extract_arg_types(signature)
+        typs = [type_from_abi({"type": x}) for x in arg_types[1:-1].split(",")]
+        inputs = list(zip(names, typs))
+        vy_ast = parse_to_ast(signature + ":\n\tpass")
+        return_type = type_from_annotation(vy_ast.body[0].returns)
+        output_signature = signature.split("->")[1].strip()
+        if not output_signature.startswith("("):
+            output_signature = f"({output_signature})"
+
```

Comment:
this can probably be replaced with a call to `ContractFunctionT.from_FunctionDef` and then grabbing the args from there